builder/openstack/step_delete_server.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package openstack
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+)
+
+type StepDeleteServer struct {
+	UseBlockStorageVolume bool
+}
+
+func (s *StepDeleteServer) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	// Only if we have a blockstorage volume we need to detach it to upload it as an image
+	if !s.UseBlockStorageVolume {
+		return multistep.ActionContinue
+	}
+
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packersdk.Ui)
+
+	instance := state.Get("instance_id").(string)
+
+	// We need the v2 compute client
+	computeClient, err := config.computeV2Client()
+	if err != nil {
+		err = fmt.Errorf("Error terminating server, may still be around: %s", err)
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	maxNumErrors := 10
+	numErrors := 0
+
+	ui.Say(fmt.Sprintf("Terminating the source server: %s ...", instance))
```

Comment:
Regarding this snippet, since it's effectively the same as the code in `step_run_source_server`, I wonder if we could factorise this so we don't repeat it twice in the codebase?

Reply:
thanks for that. I refactored it